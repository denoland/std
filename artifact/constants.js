export const PROCTYPES = {
  SELF: 'SELF',
  SPAWN: 'SPAWN',
  LOCAL: 'LOCAL',
  REMOTE: 'REMOTE',
}
