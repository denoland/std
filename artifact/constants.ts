import { IFs } from 'https://esm.sh/v135/memfs@4.6.0/lib/index.js'
import { JSONSchemaType } from 'https://esm.sh/ajv@8.12.0'
import IsolateApi from './isolate-api.ts'
export { IsolateApi }
export type { CborUint8Array } from 'https://esm.sh/v135/json-joy@9.9.1/es6/json-pack/cbor/types.d.ts?exports=CbotUint8Array'
export enum PROCTYPE {
  SERIAL = 'SERIAL',
  BRANCH = 'BRANCH',
}
export const IO_PATH = '.io.json'
export type { JSONSchemaType }

export type JsonValue = string | number | boolean | null | JsonValue[] | {
  [key: string]: JsonValue
}
export type IsolateReturn = JsonValue | void

export type IsolateFunction =
  | (() => unknown | Promise<unknown>)
  | ((...args: [Params]) => unknown | Promise<unknown>)
  | ((...args: [Params, IsolateApi]) => unknown | Promise<unknown>)

export type IsolateFunctions = {
  [key: string]: IsolateFunction
}
export type IsolateLifecycle = {
  '@@mount': (api: IsolateApi) => Promise<IsolateReturn> | IsolateReturn
  '@@unmount'?: (api: IsolateApi) => Promise<IsolateReturn> | IsolateReturn
}

export type DispatchFunctions = {
  [key: string]: (
    params?: Params,
    options?: { branch?: boolean },
  ) => unknown | Promise<unknown>
}
export type Params = Record<string, unknown>

export type IsolateApiSchema = {
  [key: string]: JSONSchemaType<any>
}
export type Isolate = {
  api: IsolateApiSchema
  functions: IsolateFunctions
  lifecycles?: IsolateLifecycle
}

export type Outcome = { result?: unknown; error?: Error }
export type IoStruct = {
  sequence: number
  requests: { [key: string]: Request }
  replies: { [key: string]: Outcome }
}
export const ENTRY_BRANCH = 'main'
/**
 * The Process Identifier used to address a specific process branch.
 */
export type PID = {
  account: string
  repository: string
  branches: string[]
}

export enum POOLABLE_TYPES {
  REPLY = 'REPLY',
  /**
   * The first action in a branch is the origin action, which will close the
   * branch once it is replied to.
   */
  DISPATCH = 'DISPATCH',
  /**
   * An external excitation.  When a reply is inserted for a pierce action, it
   * is up to the external watcher to take it from there.  In contrast when a
   * dispatch reply is received, then the execution layer needs to copy the
   * reply over to the source branch by way of a merge.
   */
  PIERCE = 'PIERCE',
}

export type Poolable = Request | Reply

export type Reply = {
  target: PID | Pierce
  sequence: number
  outcome: Outcome
  fs?: IFs
  commit?: string
}
export type Request = {
  target: PID
  source: PID | Pierce
  isolate: string
  functionName: string
  params: Params
  proctype: PROCTYPE
}
export type Pierce = {
  nonce: string
}

export enum QUEUE_TYPES {
  PULL = 'PULL',
  PUSH = 'PUSH',
  CLONE = 'CLONE',
  INIT = 'INIT',
  ISOLATE_API = 'ISOLATE_API',
  DISPATCH = 'DISPATCH',
}
export type QMessage = { nonce: string; name: string; params: Params }
export type QCallback = (
  msg: QMessage,
) => Promise<unknown> | unknown

export type QueuedDispatch = {
  type: QUEUE_TYPES.DISPATCH
  payload: {
    dispatch: Request
    sequence: number
  }
}

export type QueuedCommit = {
  type: 'COMMIT'
  payload: {
    pid: PID
    hash: string
  }
}
export enum KEYSPACES {
  POOL = 'POOL', // all pending IO actions trying to be committed
  HEADLOCK = 'HEADLOCK', // the lock on the head of a given process branch
  REPO = 'REPO', // this is the latest fs snapshot of a given process branch
  /**
   * Sequential actions that need to execute in order.  Has the form:
   *
   * [KEYSPACES.SERIAL, account, repository, ...branches, sequence] => true
   *
   * When completed, the key is deleted, which allows the next dispatch to begin
   * execution.
   */
  SERIAL = 'SERIAL',
}

export type HelpConfig = {
  model?: 'gpt-3.5-turbo-1106' | 'gpt-4-turbo-preview'
  temperature?: number
}
export type Help = {
  description?: string
  config?: HelpConfig
  runner?: string
  commands?: string[]
  instructions: string[]
  done?: string
  examples?: string[]
  tests?: string[]
}
