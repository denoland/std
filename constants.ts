import { IFs } from 'https://esm.sh/v135/memfs@4.6.0/lib/index.js'
import IsolateApi from './isolate-api.ts'
export type { IFs }
export type { IsolateApi }
export type { CborUint8Array } from 'https://esm.sh/v135/json-joy@9.9.1/es6/json-pack/cbor/types.d.ts?exports=CbotUint8Array'
export const IO_PATH = '.io.json'
import {
  IsolateApiSchema,
  IsolateReturn,
  Outcome,
  Params,
  PID,
  PierceRequest,
  PROCTYPE,
} from './api/web-client.types.ts'

export type IsolateFunction =
  | (() => unknown | Promise<unknown>)
  | ((...args: [Params]) => unknown | Promise<unknown>)
  | ((...args: [Params, IsolateApi]) => unknown | Promise<unknown>)

export type IsolateFunctions = {
  [key: string]: IsolateFunction
}
export type IsolateLifecycle = {
  '@@mount': (api: IsolateApi) => Promise<IsolateReturn> | IsolateReturn
  '@@unmount'?: (api: IsolateApi) => Promise<IsolateReturn> | IsolateReturn
}
export type Isolate = {
  api: IsolateApiSchema
  functions: IsolateFunctions
  lifecycles?: IsolateLifecycle
}
export type IoStruct = {
  sequence: number
  requests: { [key: string]: Request }
  replies: { [key: string]: Outcome }
}
export type Request = PierceRequest | SolidRequest
export type Poolable = Request | SolidReply | MergeReply
export type Reply = PierceReply | SolidReply | MergeReply
type Invocation = {
  isolate: string
  functionName: string
  params: Params
  proctype: PROCTYPE
}
/**
 * A request that has been included in a commit, therefore has a sequence number
 */
export type SolidRequest = Invocation & {
  target: PID
  source: PID
  sequence: number
}
export type EffectRequest = {
  target: PID
  /**
   * The hash of the function that was called, to ensure repeatability
   */
  fingerprint: string
  sequence: number
}
export type PierceReply = {
  ulid: string
  outcome: Outcome
}
export type SolidReply = {
  target: PID
  sequence: number
  outcome: Outcome
}
export type MergeReply = {
  target: PID
  /**
   * Where did this merge reply come from?
   */
  source: PID
  sequence: number
  outcome: Outcome
  /**
   * What is the commit that solidified this merge reply?
   */
  commit: string
}
export type IsolatePromise = {
  outcome?: Outcome
  request: SolidRequest
  resolve?: (value: unknown) => void
  reject?: (error: Error) => void
}
export type ExeResult = {
  settled?: {
    reply: SolidReply
    /**
     * Any paths that were changed on the filesystem in any way
     */
    upserts: string[]
    /**
     * Any paths that were deleted from the filesystem
     */
    deletes: string[]
  }
  pending?: { requests: SolidRequest[] }
}
export const isPierceRequest = (p: Request): p is PierceRequest => {
  return 'ulid' in p
}
export const toRunnableRequest = (request: Request, sequence: number) => {
  if (!isPierceRequest(request)) {
    return request
  }
  const { isolate, functionName, params, proctype, target } = request
  const internal: SolidRequest = {
    isolate,
    functionName,
    params,
    proctype,
    source: target,
    target,
    sequence,
  }
  return internal
}
export const isRequest = (poolable: Poolable): poolable is Request => {
  return 'proctype' in poolable
}
export const isMergeReply = (poolable: Reply): poolable is MergeReply => {
  return 'commit' in poolable
}
export const isPierceReply = (reply: Reply): reply is PierceReply => {
  return ('ulid' in reply)
}
export * from './api/web-client.types.ts'
