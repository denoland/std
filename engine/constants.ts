import IA from './isolate-api.ts'
export type { IA }
export const IO_PATH = '.io.json'
import {
  type Backchat,
  Change,
  CommitObject,
  EngineInterface,
  IsolateReturn,
  MetaPromise,
  Outcome,
  Params,
  PID,
  Pierce,
  RemoteRequest,
  Request,
  SolidRequest,
  UnsequencedRequest,
} from './api/types.ts'
import FS from '@/git/fs.ts'
import type DB from '@/db.ts'
import type Executor from '@/exe/exe.ts'
import { assert, equal } from '@utils'
import { JsonSchema7ObjectType, zodToJsonSchema } from 'zod-to-json-schema'
import { ZodObject, ZodSchema, ZodUnknown } from 'zod'

export const REPO_LOCK_TIMEOUT_MS = 5000

export type IsolateLifecycle = {
  '@@mount'?: (api: IA) => Promise<IsolateReturn> | IsolateReturn
  '@@unmount'?: (api: IA) => Promise<IsolateReturn> | IsolateReturn
}

export type Solids = {
  oid: string
  commit: CommitObject
  /** Changed files in this commit.  Empty change signals deletion. */
  changes: { [key: string]: Change }
  exe?: { request: SolidRequest; sequence: number }
  branches: number[]
  poolables: (MergeReply | RemoteRequest)[]
  deletes: { pid: PID; commit: string }[]
}
export type Branched = {
  /** The first request in the new branch */
  origin: SolidRequest
  /** The branch PID that needs to be created in /.git/refs */
  pid: PID
  /** The head of the new branch that needs to be created in /.git/refs */
  head: string
}

/**
 * Messages that go on the queue are one of three types.  Each one is an
 * operation that will result in a new commit, atomically.  Each operation is
 * able to detect when it is a duplicate task due to duplicate message delivery.
 * Each task will continue to retry until it is successful, as long as its check
 * for duplication reassures it to keep trying.
 */
export type QueueMessage = QueuePool | QueueExe | QueueBranch
export enum QueueMessageType {
  POOL = 'pool',
  EXECUTION = 'exe',
  BRANCH = 'branch',
}
export type QueuePool = {
  type: QueueMessageType.POOL
  pid: PID
}
export type QueueExe = {
  type: QueueMessageType.EXECUTION
  pid: PID
  commit: string
  sequence: number
}
export type QueueBranch = {
  type: QueueMessageType.BRANCH
  parentCommit: string
  parentPid: PID
  sequence: number
}
export const isQueuePool = (m: QueueMessage): m is QueuePool => {
  return m.type === QueueMessageType.POOL
}
export const isQueueExe = (m: QueueMessage): m is QueueExe => {
  return m.type === QueueMessageType.EXECUTION
}
export const isQueueBranch = (m: QueueMessage): m is QueueBranch => {
  return m.type === QueueMessageType.BRANCH
}
export const isChildOf = (child: PID, parent: PID) => {
  const childParent = { ...child, branches: child.branches.slice(0, -1) }
  return equal(childParent, parent)
}
export const isBaseRepo = (pid: PID) => pid.branches.length === 1
