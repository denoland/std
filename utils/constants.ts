// Copyright 2023-2024 the Deno authors. All rights reserved. MIT license.
export const SITE_NAME = "Deno Hunt (Beta)";
export const SITE_DESCRIPTION = "Discover new Deno projects. Share your own.";
