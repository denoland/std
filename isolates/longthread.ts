import { assert, Debug, equal } from '@utils'
import {
  Agent,
  CompletionMessage,
  getThreadPath,
  IA,
  pidSchema,
  print,
  Thread,
  ToApiType,
} from '@/constants.ts'
import { ToolMessage } from '@/api/zod.ts'
import { Functions } from '@/constants.ts'
import { load } from './utils/load-agent.ts'
import { executeTools } from './utils/ai-execute-tools.ts'
import { z } from 'zod'
import * as completions from '@/isolates/ai-completions.ts'
import * as agents from '@/isolates/agents.ts'
const log = Debug('AI:longthread')

export const parameters = {
  start: z.object({}).describe('start a new thread for the given agent'),
  run: z.object({
    /** Path to the agent to instantiate */
    path: z.string(),
    content: z.string(),
    actorId: z.string(),
  }),
  switchboard: z.object({
    content: z.string(),
    actorId: z.string(),
  }),
  drone: z.object({
    /** Path to the agent to instantiate */
    path: z.string(),
    content: z.string(),
    actorId: z.string(),
    /** tool names to stop on */
    stopOnTools: z.array(z.string()),
  }),
  changeRemote: z.object({
    remote: pidSchema.optional(),
  }),
}
export const returns = {
  start: z.void(),
  run: z.void(),
  /** If the tools reqeuested a new thread to be formed, or a change to the
   * backchat target thread */
  switchboard: z.object({
    newThread: z.boolean().optional(),
    changeThread: pidSchema.optional(),
  }).refine((v) => !(v.newThread && v.changeThread), {
    message: 'expected newThread or changeThread exclusively',
  }),
  drone: z.union([
    z.object({ functionName: z.string(), args: z.record(z.unknown()) }),
    z.undefined(),
  ]),
  changeRemote: z.void(),
}

export type Api = ToApiType<typeof parameters, typeof returns>

export const functions: Functions<Api> = {
  start: async (_, api) => {
    const threadPath = getThreadPath(api.pid)
    log('start', threadPath, print(api.pid))
    assert(!await api.exists(threadPath), `thread exists: ${threadPath}`)
    const thread: Thread = {
      messageOffset: 0,
      messages: [],
      toolCommits: {},
      stateboards: [],
      focusedFiles: [],
    }
    api.writeJSON(threadPath, thread)
  },
  run: async ({ path, content, actorId }, api) => {
    // TODO sniff actorId from the action source
    log('run', { path, content, actorId })
    const threadPath = getThreadPath(api.pid)
    let thread = await api.readThread(threadPath)

    thread.messages.push({ name: actorId, role: 'user', content })
    api.writeJSON(threadPath, thread)
    await loop(path, api, [])
    thread = await api.readThread(threadPath)
    const last = thread.messages[thread.messages.length - 1]
    assert(last.role === 'assistant', 'not assistant: ' + last.role)
    assert(typeof last.content === 'string', 'expected string content')
  },
  switchboard: async ({ content, actorId }, api) => {
    // TODO handle remote threadIds with symlinks in the threads dir

    const path = `agents/switchboard.md`
    // TODO change this to be a drone call
    // verify the switchboard has the right tool loaded
    // TODO verify stopOnTool function returns null

    const overrides = await load(path, api)
    overrides.config.parallel_tool_calls = false
    overrides.config.tool_choice = 'required'
    overrides.commands.push('agents:switch')

    const threadPath = getThreadPath(api.pid)
    const thread = await api.readJSON<Thread>(threadPath)
    thread.messages.push({ name: actorId, role: 'user', content })
    api.writeJSON(threadPath, thread)

    const switchResult = await loop(path, api, ['agents_switch'], overrides)
    assert(switchResult, 'expected switchboard result')
    assert(switchResult.functionName === 'agents_switch', 'not agents_switch')
    const { path: next } = agents.parameters.switch.parse(switchResult.args)

    const stopOnTools = []
    if (next === `agents/backchat.md`) {
      stopOnTools.push(
        'backchat_newThreadSignal',
        'backchat_changeThreadSignal',
      )
    }

    const agentResult = await loop(next, api, stopOnTools)
    const result: z.infer<typeof returns.switchboard> = {}
    if (agentResult) {
      const { functionName } = agentResult
      if (functionName === 'backchat_newThreadSignal') {
        result.newThread = true
      } else if (functionName === 'backchat_changeThreadSignal') {
        result.changeThread = api.pid
      }
    }
    // TODO verify the agent can actually end somehow
    return result
  },
  drone: async ({ path, content, actorId, stopOnTools }, api) => {
    // TODO verify stopOnTools are all present in the agent and have the
    // expected return value of null
    log('drone', path, content, actorId, stopOnTools)
    await functions.start({}, api)

    const threadPath = getThreadPath(api.pid)
    const thread = await api.readJSON<Thread>(threadPath)
    thread.messages.push({ name: actorId, role: 'user', content })
    api.writeJSON(threadPath, thread)
    return loop(path, api, stopOnTools)
  },
  changeRemote: async ({ remote }, api) => {
    const threadPath = getThreadPath(api.pid)
    const thread = await api.readThread(threadPath)

    if (equal(thread.remote, remote)) {
      return
    }
    const id = 'call_keVDORgd16XuNBh0PMk5leKP'
    thread.messages.push({
      role: 'assistant',
      tool_calls: [
        {
          type: 'function',
          function: {
            name: 'SYSTEM_changeRemote',
            arguments: JSON.stringify(remote ? { remote } : {}),
          },
          id,
        },
      ],
    }, { role: 'tool', content: 'null', tool_call_id: id })

    if (remote) {
      const opts = { target: remote }
      await api.readThread(getThreadPath(remote), opts)
      thread.remote = remote
    } else {
      delete thread.remote
    }
    api.writeJSON(threadPath, thread)
  },
}

const loop = async (
  path: string,
  api: IA,
  stopOnTools: string[],
  overrides?: Partial<Agent>,
) => {
  const threadPath = getThreadPath(api.pid)
  const { complete } = await api.actions<completions.Api>('ai-completions')
  const HARD_STOP = 20
  let count = 0
  stopOnTools = addDefaults(stopOnTools)

  while (!await isDone(threadPath, api, stopOnTools) && count++ < HARD_STOP) {
    await complete({ path, overrides })
    if (await isDone(threadPath, api)) {
      break
    }
    // TODO check tool returns are checked against schema
    await executeTools(threadPath, api, stopOnTools, overrides)
  }
  if (count >= HARD_STOP) {
    throw new Error('LONGTHREAD hard stop after: ' + HARD_STOP + ' loops')
  }

  const thread = await api.readThread(threadPath)
  const last = thread.messages[thread.messages.length - 1]
  if (last.role === 'tool') {
    const assistant = thread.messages[thread.messages.length - 2]
    assert(assistant.role === 'assistant', 'not assistant: ' + last.role)
    assert(assistant.tool_calls?.length === 1, 'expected one tool call')

    const args = JSON.parse(assistant.tool_calls[0].function.arguments)
    const functionName = assistant.tool_calls[0].function.name
    log('tool call', functionName, args)
    return { functionName, args }
  }
}

const isDone = async (threadPath: string, api: IA, stopOnTools?: string[]) => {
  const thread = await api.readJSON<Thread>(threadPath)
  const last = thread.messages[thread.messages.length - 1]
  if (!last) {
    return false
  }
  if (last.role === 'user' || last.role === 'assistant') {
    log(`${last.role}:${last.name}:`, last.content)
  }
  if (last.role === 'tool') {
    if (stopOnTools) {
      const prior = thread.messages[thread.messages.length - 2]
      if (isStopOnTool(prior, last, stopOnTools)) {
        return true
      }
    }
  }
  if (last.role !== 'assistant') {
    return false
  }
  if ('tool_calls' in last) {
    return false
  }
  return true
}

const isStopOnTool = (
  prior: CompletionMessage,
  tool: ToolMessage,
  stopOnTools: string[],
) => {
  if (prior.role !== 'assistant') {
    return false
  }
  if (!prior.tool_calls) {
    return false
  }
  if (prior.tool_calls.length !== 1) {
    return false
  }
  if (!stopOnTools.includes(prior.tool_calls[0].function.name)) {
    return false
  }
  if (tool.content !== 'null') {
    return false
  }
  return true
}
const addDefaults = (stopOnTools: string[]) => {
  const result = [...stopOnTools]
  if (!result.includes('utils_resolve')) {
    result.push('utils_resolve')
  }
  if (!result.includes('utils_reject')) {
    result.push('utils_reject')
  }
  return result
}
