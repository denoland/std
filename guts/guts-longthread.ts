import { assert, expect, log } from '@utils'
import { CradleMaker, print } from '@/constants.ts'
import { Api } from '../openai/longthread.ts'

export default (cradleMaker: CradleMaker) => {
  const prefix = 'longthread: '
  Deno.test(prefix + 'basic', async (t) => {
    await using cradle = await cradleMaker(t, import.meta.url)
    const { backchat } = cradle
    log('pid', print(backchat.pid))

    await t.step('prompt', async () => {
      const { start, run } = await backchat.actions<Api>('longthread')
      await start({})
      await run({
        path: 'agents/agent-fixture.md',
        content: 'say "Hello"',
        actorId: 'ai',
      })
      const thread = await backchat.readThread(backchat.pid)
      const assistant = thread.messages.pop()
      assert(assistant)
      expect(assistant.content).toContain('Hello')
    })
    log('stopping')
    // TODO test using the backchat thread function directly
  })

  Deno.test(prefix + 'chat', async (t) => {
    await using cradle = await cradleMaker(t, import.meta.url)
    const { backchat } = cradle

    const { start, run } = await backchat.actions<Api>('longthread')
    await start({})

    await t.step('say the word "hello"', async () => {
      await run({
        path: 'agents/agent-fixture.md',
        content: 'say the word: "hello" verbatim without calling any functions',
        actorId: 'ai',
      })
      const thread = await backchat.readThread(backchat.pid)
      expect(thread.messages).toHaveLength(2)

      const assistant = thread.messages.pop()
      assert(assistant)
      expect(assistant.content).toBe('hello')
    })

    await t.step('what is your name ?', async () => {
      await run({
        content: 'what is your name ?',
        path: 'agents/agent-fixture.md',
        actorId: 'ai',
      })
      const thread = await backchat.readThread(backchat.pid)
      expect(thread.messages).toHaveLength(4)
      const assistant = thread.messages.pop()
      assert(assistant)

      const expected = `Assistant`
      expect(assistant.content).toContain(expected)
    })

    await t.step('repeat your last', async () => {
      await run({
        content: 'repeat your last, without calling any functions',
        path: 'agents/agent-fixture.md',
        actorId: 'ai',
      })
      const thread = await backchat.readThread(backchat.pid)
      expect(thread.messages).toHaveLength(6)
      const assistant = thread.messages.pop()
      assert(assistant)

      const expected = `Assistant`
      expect(assistant.content).toContain(expected)
    })
  })
}

// add a user to the system
// add an anonymous user before they log in
// let the user log in to the system
