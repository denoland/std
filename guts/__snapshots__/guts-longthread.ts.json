{
  "longthread: basic": {
    "recordings": {
      "rep_E8R/dreamcatcher-tech/HAL:main/act_H2R/bac_VWV": [
        {
          "request": {
            "model": "gpt-4o-mini",
            "tool_choice": "auto",
            "parallel_tool_calls": true,
            "messages": [
              {
                "content": "say \"Hello\"",
                "role": "user",
                "name": "ai"
              },
              {
                "content": "Your name is 'Assistant'\n\nALWAYS be as brief as possible",
                "role": "system",
                "name": "agents_agent-fixture_md"
              }
            ],
            "seed": 1337,
            "tools": [
              {
                "type": "function",
                "function": {
                  "name": "io-fixture_local",
                  "description": "ping locally",
                  "parameters": {
                    "type": "object",
                    "properties": {},
                    "additionalProperties": false
                  },
                  "strict": false
                }
              },
              {
                "type": "function",
                "function": {
                  "name": "io-fixture_error",
                  "description": "throw an error",
                  "parameters": {
                    "type": "object",
                    "properties": {
                      "message": {
                        "type": "string"
                      }
                    },
                    "required": [
                      "message"
                    ],
                    "additionalProperties": false
                  },
                  "strict": false
                }
              }
            ]
          },
          "reply": {
            "assistant": {
              "role": "assistant",
              "content": "Hello!",
              "name": "agents/agent-fixture.md",
              "refusal": null
            },
            "stats": {
              "created": 1729382538,
              "model": "gpt-4o-mini-2024-07-18",
              "system_fingerprint": "fp_482c22a7bc",
              "duration": 577,
              "openAiProcessingMs": 339,
              "usage": {
                "completion_tokens": 3,
                "prompt_tokens": 85,
                "total_tokens": 88,
                "completion_tokens_details": {
                  "reasoning_tokens": 0
                },
                "prompt_tokens_details": {
                  "cached_tokens": 0
                }
              }
            }
          }
        }
      ]
    }
  },
  "longthread: chat": {
    "recordings": {
      "rep_E8R/dreamcatcher-tech/HAL:main/act_H2R/bac_VWV": [
        {
          "request": {
            "model": "gpt-4o-mini",
            "tool_choice": "auto",
            "parallel_tool_calls": true,
            "messages": [
              {
                "content": "say the word: \"hello\" verbatim without calling any functions",
                "role": "user",
                "name": "ai"
              },
              {
                "content": "Your name is 'Assistant'\n\nALWAYS be as brief as possible",
                "role": "system",
                "name": "agents_agent-fixture_md"
              }
            ],
            "seed": 1337,
            "tools": [
              {
                "type": "function",
                "function": {
                  "name": "io-fixture_local",
                  "description": "ping locally",
                  "parameters": {
                    "type": "object",
                    "properties": {},
                    "additionalProperties": false
                  },
                  "strict": false
                }
              },
              {
                "type": "function",
                "function": {
                  "name": "io-fixture_error",
                  "description": "throw an error",
                  "parameters": {
                    "type": "object",
                    "properties": {
                      "message": {
                        "type": "string"
                      }
                    },
                    "required": [
                      "message"
                    ],
                    "additionalProperties": false
                  },
                  "strict": false
                }
              }
            ]
          },
          "reply": {
            "assistant": {
              "role": "assistant",
              "content": "hello",
              "name": "agents/agent-fixture.md",
              "refusal": null
            },
            "stats": {
              "created": 1729382540,
              "model": "gpt-4o-mini-2024-07-18",
              "system_fingerprint": "fp_482c22a7bc",
              "duration": 401,
              "openAiProcessingMs": 187,
              "usage": {
                "completion_tokens": 2,
                "prompt_tokens": 94,
                "total_tokens": 96,
                "completion_tokens_details": {
                  "reasoning_tokens": 0
                },
                "prompt_tokens_details": {
                  "cached_tokens": 0
                }
              }
            }
          }
        },
        {
          "request": {
            "model": "gpt-4o-mini",
            "tool_choice": "auto",
            "parallel_tool_calls": true,
            "messages": [
              {
                "content": "say the word: \"hello\" verbatim without calling any functions",
                "role": "user",
                "name": "ai"
              },
              {
                "role": "assistant",
                "content": "hello",
                "name": "agents_agent-fixture_md",
                "refusal": null
              },
              {
                "content": "what is your name ?",
                "role": "user",
                "name": "ai"
              },
              {
                "content": "Your name is 'Assistant'\n\nALWAYS be as brief as possible",
                "role": "system",
                "name": "agents_agent-fixture_md"
              }
            ],
            "seed": 1337,
            "tools": [
              {
                "type": "function",
                "function": {
                  "name": "io-fixture_local",
                  "description": "ping locally",
                  "parameters": {
                    "type": "object",
                    "properties": {},
                    "additionalProperties": false
                  },
                  "strict": false
                }
              },
              {
                "type": "function",
                "function": {
                  "name": "io-fixture_error",
                  "description": "throw an error",
                  "parameters": {
                    "type": "object",
                    "properties": {
                      "message": {
                        "type": "string"
                      }
                    },
                    "required": [
                      "message"
                    ],
                    "additionalProperties": false
                  },
                  "strict": false
                }
              }
            ]
          },
          "reply": {
            "assistant": {
              "role": "assistant",
              "content": "Assistant.",
              "name": "agents/agent-fixture.md",
              "refusal": null
            },
            "stats": {
              "created": 1729382541,
              "model": "gpt-4o-mini-2024-07-18",
              "system_fingerprint": "fp_482c22a7bc",
              "duration": 391,
              "openAiProcessingMs": 183,
              "usage": {
                "completion_tokens": 3,
                "prompt_tokens": 116,
                "total_tokens": 119,
                "completion_tokens_details": {
                  "reasoning_tokens": 0
                },
                "prompt_tokens_details": {
                  "cached_tokens": 0
                }
              }
            }
          }
        },
        {
          "request": {
            "model": "gpt-4o-mini",
            "tool_choice": "auto",
            "parallel_tool_calls": true,
            "messages": [
              {
                "content": "say the word: \"hello\" verbatim without calling any functions",
                "role": "user",
                "name": "ai"
              },
              {
                "role": "assistant",
                "content": "hello",
                "name": "agents_agent-fixture_md",
                "refusal": null
              },
              {
                "content": "what is your name ?",
                "role": "user",
                "name": "ai"
              },
              {
                "role": "assistant",
                "content": "Assistant.",
                "name": "agents_agent-fixture_md",
                "refusal": null
              },
              {
                "content": "repeat your last, without calling any functions",
                "role": "user",
                "name": "ai"
              },
              {
                "content": "Your name is 'Assistant'\n\nALWAYS be as brief as possible",
                "role": "system",
                "name": "agents_agent-fixture_md"
              }
            ],
            "seed": 1337,
            "tools": [
              {
                "type": "function",
                "function": {
                  "name": "io-fixture_local",
                  "description": "ping locally",
                  "parameters": {
                    "type": "object",
                    "properties": {},
                    "additionalProperties": false
                  },
                  "strict": false
                }
              },
              {
                "type": "function",
                "function": {
                  "name": "io-fixture_error",
                  "description": "throw an error",
                  "parameters": {
                    "type": "object",
                    "properties": {
                      "message": {
                        "type": "string"
                      }
                    },
                    "required": [
                      "message"
                    ],
                    "additionalProperties": false
                  },
                  "strict": false
                }
              }
            ]
          },
          "reply": {
            "assistant": {
              "role": "assistant",
              "content": "Assistant.",
              "name": "agents/agent-fixture.md",
              "refusal": null
            },
            "stats": {
              "created": 1729382542,
              "model": "gpt-4o-mini-2024-07-18",
              "system_fingerprint": "fp_482c22a7bc",
              "duration": 467,
              "openAiProcessingMs": 257,
              "usage": {
                "completion_tokens": 3,
                "prompt_tokens": 142,
                "total_tokens": 145,
                "completion_tokens_details": {
                  "reasoning_tokens": 0
                },
                "prompt_tokens_details": {
                  "cached_tokens": 0
                }
              }
            }
          }
        }
      ]
    }
  }
}
